### PR TITLE
Feat add services interface in shared

### PR DESCRIPTION
**Pull request by LudovicSterlin**
## Description

Provide types and interfaces for services in API extension context by providing a `ServicesOverview` to replace the `any` type in `ApiExtensionContext`:

https://github.com/directus/directus/blob/1fedc5bdfd7ef7f731ce0bddd6b2e1c50d69f120/packages/shared/src/types/extensions.ts#L132-L134

Which would then look like this and allow your IDE to inform you on all the services available when developing extensions.
```ts
export type ApiExtensionContext = {
	services: ServicesOverview;
	[...]
};
```

This is done by moving all the types necessary for the services interfaces in `@directus/shared` and implementing all the services interfaces in a new folder of shared package `services`

Fixes #12563, also probably a good first step to improve [directus/docs/#20](https://github.com/directus/docs/issues/20)

Note: no test were written, and the doc is not updated yet because some stuff need discussion in this PR

## Type of Change

- [ ] Bugfix
- [x] Improvement
- [ ] New Feature
- [x] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [x] New / updated tests are included
- [x] All tests are passing locally
- [x] Performed a self-review of the submitted code

If adding a new feature:

- [ ] Documentation was added/updated. PR:

### CODE REVIEW

File: packages/shared/src/services/assets.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import { Accountability, TransformationParams, TransformationPreset } from '../types';
+import { AuthorizationService } from './authorization';
+export declare interface AssetsService {
+	knex: Knex;
+	accountability: Accountability | null;
+	authorizationService: AuthorizationService;
+	getAsset(
+		id: string,
+		transformation: TransformationParams | TransformationPreset,
+		range?: any // should be Range from '@directus/drive';
```
Comment (LudovicSterlin) [inline packages/shared/src/services/assets.ts:11]: Trying to implement the services interfaces as suggested in this comment:
> [...] The best we can do is probaly still to create dedicated interfaces for each service inside shared and let the actual service classes implement those. [...]
> _Originally posted by @nickrum in https://github.com/directus/directus/discussions/12563#discussioncomment-3447115_

 One of the issue I ran into was that some services function parameters types come from `@directus/drive` or `@directus/schema` package.

For example, this is the case here with `Range` and `StatResponse`.

File: packages/shared/src/services/collections.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	collection: string;
+	fields?: RawField[];
+	schema?: Partial<Table> | null;
+	meta?: Partial<CollectionMeta> | null;
+};
+export declare interface CollectionsService {
+	knex: Knex;
+	helpers: Helpers;
+	accountability: Accountability | null;
+	schemaInspector: any; // should be ReturnType<typeof SchemaInspector>; with SchemaInspector from '@directus/schema';
```
Comment (LudovicSterlin) [inline packages/shared/src/services/collections.ts:23]: Same problem here with `SchemaInspector` from `@directus/schema`
The dependence to  `@directus/schema` was also present in :
- [fields.ts](https://github.com/directus/directus/blob/d7f27830e43015ef111c297172ab46ebca29c97b/packages/shared/src/services/fields.ts)
- [relations.ts](https://github.com/directus/directus/blob/d7f27830e43015ef111c297172ab46ebca29c97b/packages/shared/src/services/relations.ts)

File: packages/shared/src/types/helpers.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+// would need to move all the types from api/src/database/helpers to type this one better
+export declare type Helpers = {
+	date: any;
+	st: any;
+	schema: any;
+};
```
Comment (LudovicSterlin) [inline packages/shared/src/types/helpers.ts:6]: Here moving all the type to properly type `Helpers` seems too much of a refactoring, that's why I just make a new type with only the structure and not the type of the properties inside. 

The 'real' type is in the api, here https://github.com/directus/directus/blob/d7f27830e43015ef111c297172ab46ebca29c97b/api/src/database/helpers/index.ts#L25

This duplicate `type Helpers` is used in `RelationsService` https://github.com/directus/directus/blob/d7f27830e43015ef111c297172ab46ebca29c97b/packages/shared/src/services/relations.ts#L14)

**Comment by rijkvanzanten** [comment]
Closing this in favor of the newly created `@directus/types`, which is the new home for these shared types ✨

**Pull request closed without merging**